### DISCUSSION TITLE
Leaderboard for custom environments

### DISCUSSION THREAD

**Discussion by LaurensNeinders** [trackmania-rl/tmrl#33] [Ideas]
Hi,
I made an AI for Trackmania using tmrl and didn't use one of the given environments, but made a custom one. I managed a fastest time of 38.398 seconds and and average just under 40 seconds. 
What do you think about having a separate leaderboard with more flexible rules (like custom environments)?

**Comment by yannbouteiller** [comment]
Hello, this is totally feasible, there is no fixed rule for the competition at the moment.

We are working toward proposing this competition to a robot conference, at which point we will need a set of well-defined rules that interest everyone. So it would be good to hear your opinion about the default environments and why you chose to design your own, too?

**Reply by LaurensNeinders** [reply]
You might know Sophy AI by Sony which is very fast in Gran Turismo.
I compared the lidar environment to this and found that one of the main differences was that Sophy gets 6 seconds of track in advance as input for the AI. 
I thought that the main problem why an AI in the lidar and full environment might not be able to drive very quick at the moment is because they cannot "see" that far.  the lidar-sensors for example can only see the track that is right in front of it. It doesn't know how the corner after this one will look, so it cannot anticipate what the best way of driving through it is. 
The environment I made has as input a representation of where the walls are in the upcoming piece of track. 

![lidar vs trackinfo_sm](https://user-images.githubusercontent.com/66493565/214573656-3a16dcb2-f628-44cd-8fda-7eef6a1e8940.png)

**Reply by NDR008** [reply]
@LaurensNeinders : Interesting, actually I recently referenced Sophy on here, and am actually trying to do something on Gran Turismo for my dissertation.
Regarding Sophy's ability to see the track 6 seconds ahead - how do you know this?

**Reply by LaurensNeinders** [reply]
I found it on the blog about Sophy: 
[https://ai.sony/blog/blog-019/#:~:text=%2C%20up%20to%20six%20seconds%20from%20its%20current%20position](url)

**Comment by LaurensNeinders** [comment]
For some more info, I could send you the draft version of the report I am writing about it via discord

**Reply by yannbouteiller** [reply]
Oh I see! Pretty cool project.

I think it indeed makes sense to have a leaderboard for this kind of alternative environments so people can see what really works in the current SOTA. 

Not sure whether we can easily integrate such environments in the repo because I expect people will need their own custom OpenPlanet plugins for this, but as long as we can reproduce the results it is definitely a good idea to have a leaderboard showcasing projects like this.

**Reply by LaurensNeinders** [reply]
Hey Yann, 

For the report I would like to know what is the current fastest time by the lidar-AI on the tmrl-test track?

**Reply by yannbouteiller** [reply]
I think the "cheat" version that we used for the Underscore_ show (the one that fed the track progress % along with the LIDAR to the MLP) had a best time of 43.5s or so, and a mean time of ~45s on the `tmrl-test` track

**Comment by yannbouteiller** [comment]
Just to let you know, I have made a [leaderboard](https://github.com/trackmania-rl/tmrl/blob/master/readme/competition.md#freestyle-off-competition-unicorn) for your project :p

**Reply by LaurensNeinders** [reply]
That sounds great, I will notify you when I have the repository ready

**Comment by eliegoudout** [comment]
Hello,
I'm very interested in this project, even though I currently cannot free enough time to contribute or even play around with it.

Regarding the "real world"-style leaderboard, I think it can be really really interesting to differentiate further into 2 categories: _human-like_ and _robot-like_.

What I mean by this is that humans are limited in many ways unknown to robots sensors, and I genuinely think that working on creating a "human-limited" agent can in fact help us better understand how humans learn, and progress towards GAI.

For example, here are some limitations I would include in a human-like environment:
- Hand limitations: Impossible to have flickering inputs (we can't tap "tight/left" 10 times a second),
- Sight limitations: our sight moves somewhat "slowly", and the more we see, the more blurry it gets (whereas the more focused is our sight, the clearer we see). Furthermore, it takes some amount of time to switch from focusing on a tiny region to looking at a large area.
- Reaction limitation: our brain has several frames of "input lag" if you will. It can be added artificially.

The sight limitations are the more tricky ones, but also the most important in my opinion. For example, no human can constantly Know his or her exact speed at all time by driving normally and constantly evaluating the bottom right corner speedometer.

I've been thinking about this for several years now, but never got around digging further: I think human features used during driving in essence come down to switching between tiny focused important areas and large blurry areas, constantly trying to find the "spot of interest" in the images.


I'm kind of throwing that out there, pretty much without any concrete solution and not a lot of structure, but maybe this can be a topic for discussion depending on the goal of this project.


Sincerely.
Èlie 

Edit: I realize that some limitations are naturally on the side of the environment (_e.g._ sight lag), but some are less natural to put into the environment. It would somehow require the environment to communicate with the agent and remember informations about its previous states to limit its next observation.

**Reply by NDR008** [reply]
We have similar interests! Dr Joscha Bach is often using NN to understand humans - I love watching his talks.
The SophyAI method of seeing 6 seconds ahead initially felt like "cheating" but thinking about it, it would be cheating for a true robotics in an unknown environment, but feels fair for a racing challenge.
A human race-car driver would have a reliable model of what the next 2 corners are like (from memory).
In fact, Lidar technically tries to see ahead, just cannot see round corners.

I guess Sophy would be cheating if it can also see where the competitor cars are in places were a human could not.
